text/000-rustdoc-jump-to-definition.md

```diff
@@ -0,0 +1,414 @@
+
+## Link generation
+
+There are two kinds of links:
+
+ * Jump to documentation.
+ * Jump to definition.
+
+### Jump to documentation
+
+All items with a documentation page will have a link to it (like methods, trait/impl associated items, etc). However, fields, variants and impl blocks won't (explications below).
+
+The link will be generated on their ident/name where they are declared, not when they are in use. For example:
+
+```rust
+pub struct Foo;
+
+let x = Foo;
+```
+
+On `Foo` definition (`pub struct Foo`), we will generate a link to its documentation page whereas on its usage (`x = Foo`), we will generate a link to jump to its definition.
+
+This is why generating a link for the impl blocks would be complicated: on `impl Foo`, we will generate a "jump to definition" link on `Foo`, leaving no space to generate a link to the impl documentation.
+
+For fields and variants, it's simply to limit the quantity of links generated: you can simply click on their parent item to get to their documentation page.
+
+### Jump to definition
+
+A small preamble before listing the items we should generate links for: if an item which will get a link is preceded by a path, each part of the path will be a link to its target. For example:
+
+```rust
+let x: b::c::D; // `b` is a link to `b`, `c` is a link to `b::c`, `D` is a link to `b::c::D`.
+foo::some_fn(); // `foo` is a link to `foo` and `some_fn` is the link to `foo::some_fn`.
+```
+
+Anything not specifically mentioned won't get a link generated for itself.
+
+#### Local variables
+
+Local variables will get a link to their definition (not their type):
+
+```rust
+fn f(mut a: String) {
+    a.push_str("something"); // `a` get a link just like `push_str`!
+}
+```
+
+The reasons about supporting local variables are as follow:
+
+ * We want to linkify global variables, because they may be far away or in another file.
+ * If you're looking at a variable reference in a function, you don't know whether it's a local variable or a global one (or one from an enclosing scope). By linkifying variables we make it easy for people to find out the answer. If we linkify only some variables (the global ones), it will appear weird that some are linkified and some are not. So for coherency, it will be done for all local variables.
+
+#### Primitive types
+
+No primitive type will get a link.
+
+#### Fields
+
+They will not get a link either.
+
+#### Variants
+
+They will get a link, but only on their usage (as already specified in the "jump to documentation" section):
+
+```rust
+enum Enum {
+    Variant, // No link here.
+}
+
+// `Enum` links to `Enum` and `Variant` is the link to `Enum::Variant` definition.
+let x = Enum::Variant;
+```
+
+#### Imports
+
+Imports will have links as well, following the [Paths](#Paths) rules.
+
+#### Types
+
+Whenever a type appears, we will link to its definition. For example:
```

Comment:
```suggestion
Whenever we encounter a type path that refer to types defined by an item (this excludes type parameters), we will link to its definition. For example:
```

Minor nit: To be technically more correct, you want to refer to *type paths* here since arbitrary *types* includes `dyn Thing`, `*const ()`, `[i32]` and the like.
The phrasing of *refer to types defined by an item* can probably be improved. As the example shows, we don't want to generate links for the type parameters `F` and `T`. So let's mention it in prose.

Reply:
:+1: